src/crimson/osd/backfill_state.h

```diff
@@ -29,61 +32,191 @@ struct BackfillState {
     BackfillInterval result;
   };
 
-  struct Flushed : sc::event<Flushed> {
+  struct ObjectPushed : sc::event<ObjectPushed> {
+    // TODO: implement replica management; I don't want to follow
+    // current convention where the backend layer is responsible
+    // for tracking replicas.
+    pg_shard_t replica; // on which replica
+    hobject_t object;
+    pg_stat_t stat;
   };
 
   struct Trigerred : sc::event<Trigerred> {
   };
 
+private:
+  // internal events
+  struct RequestPrimaryScanning : sc::event<RequestPrimaryScanning> {
+  };
+
+  struct RequestReplicasScanning : sc::event<RequestReplicasScanning> {
+  };
+
+  struct RequestWaiting : sc::event<RequestWaiting> {
+  };
+
+  struct RequestDone : sc::event<RequestDone> {
+  };
+
+public:
+
   struct Initial;
   struct Enqueuing;
+  struct PrimaryScanning;
+  struct ReplicasScanning;
+  struct Waiting;
+
+  struct BackfillMachine : sc::state_machine<BackfillMachine, Initial> {
+    BackfillState& backfill_state;
+    BackfillListener& backfill_listener;
+    PeeringFacade& peering_state;
+    PGFacade& pg;
+  };
 
-  class BackfillMachine : public sc::state_machine<BackfillMachine, Initial> {
+private:
+  template <class S>
+  struct StateHelper {
+    BackfillState& bs() {
+      return static_cast<S*>(this) \
+        ->template context<BackfillMachine>().backfill_state;
+    }
+    BackfillListener& ls() {
+      return static_cast<S*>(this) \
+        ->template context<BackfillMachine>().backfill_listener;
+    }
+    PeeringFacade& ps() {
+      return static_cast<S*>(this) \
+        ->template context<BackfillMachine>().peering_state;
+    }
+    PGFacade& pg() {
+      return static_cast<S*>(this)->template context<BackfillMachine>().pg;
+    }
+
+    const PeeringFacade& ps() const {
+      return static_cast<const S*>(this) \
+        ->template context<BackfillMachine>().peering_state;
+    }
+    const BackfillState& bs() const {
+      return static_cast<const S*>(this) \
+        ->template context<BackfillMachine>().backfill_state;
+    }
   };
 
+public:
+
   // states
-  struct Crashed : sc::state<Crashed, BackfillMachine>, NamedState {
```

Comment:
Hmm, this could be implemented using a class with debugs in both ctor and dtor + a macro to avoid RTTI and the potentially ugly outcome of `typeid(T).name()`.

Reply:
Refactored with [`boost::typeindex::type_id<T>().pretty_name()`](https://www.boost.org/doc/libs/1_66_0/doc/html/boost_typeindex/getting_started.html).